felix/dataplane/linux/qos_controls.go

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2025 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package intdataplane
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/projectcalico/calico/felix/dataplane/linux/bwp"
+	"github.com/projectcalico/calico/felix/proto"
+)
+
+func (m *endpointManager) maybeUpdateQoSBandwidth(old, new *proto.WorkloadEndpoint) error {
+	var errs []error
+	var oldIngressBw, oldEgressBw, newIngressBw, newEgressBw int64
+	var oldIngressBurst, oldEgressBurst, newIngressBurst, newEgressBurst int64
+	if old != nil && old.QosControls != nil {
+		oldIngressBw = old.QosControls.IngressBandwidth
+		oldEgressBw = old.QosControls.EgressBandwidth
+		oldIngressBurst = old.QosControls.IngressBurst
+		oldEgressBurst = old.QosControls.EgressBurst
+	}
+
+	if new != nil && new.QosControls != nil {
+		newIngressBw = new.QosControls.IngressBandwidth
+		newEgressBw = new.QosControls.EgressBandwidth
+		newIngressBurst = new.QosControls.IngressBurst
+		newEgressBurst = new.QosControls.EgressBurst
+	}
+
+	if oldIngressBw != newIngressBw || oldIngressBurst != newIngressBurst {
+		if oldIngressBw != 0 {
```

Comment:
This implies that it doesn't make sense to have `xyzBurst != 0` if `xyzBandwidth` is 0.  And that `xyzBurst` values are ignored when `xyzBandwidth` is 0.  Is that correct?

Reply:
Yes, that is correct, since `XYZburst` is optional (and defaulted when only `XYZbandwidth` is provided). But that logic ended up changing quite a bit